ArduCopter/toy_mode.cpp

```diff
@@ -416,6 +416,42 @@ void ToyMode::update()
     } else {
         throttle_high_counter = 0;
     }
+    
+    /*
+    * attempt to improve the setting of home location if it takes off when EKF is not happy
```

Comment:
I don't get what you mean? - do you mean we have to create a exit mechanism so that it does not go into this code every time ToyMode::update() is called?.

Reply:
Something along the lines of:

```
void ToyMode::set_home_location_from_gps()
{
    if (!copter.motors->armed()) {
       // only reset home once the vehicle is armed (explain why here!)
       return;
    }
    if (copter.position_ok()) {
       // EKF will set home
       return;
    }

    if (copter.gps.status() < AP_GPS::GPS_OK_FIX_3D) {
      return;
    }
    if (toy_horizontal_acc >15.0f) {
      // trust the GPS's guess of its accuracy
      return;
    }

.
.
.

}

```

And add a call to this function where your code currently is.

This has the advantage of not ending up nested 4-deep...